### PR TITLE
Generalize s3-backup

### LINKED ISSUE #2: Could make it compatible with other S3 providers?
leaving a custom URL option would be great

### PR DESCRIPTION
**Pull request by hobbypunk90**
Closes #2 :grin: 
I run my changes with scaleways object storage :relaxed:

### CODE REVIEW

**Comment by thomasfr** [comment]
Thanks for the pull request. Very much appreciated!! I will take a look into it this weekend

**Comment by hobbypunk90** [comment]
@thomasfr i hope it's okay, that i've added also an option to set the max_concurrent_requests param for aws.

**Comment by hobbypunk90** [comment]
@thomasfr do you had time for a look? :)

File: s3-backup/run.sh [hunk 1]
```diff
@@ -1,26 +1,38 @@
...
+aws configure set plugins.endpoint awscli_plugin_endpoint
+if bashio::config.has_value "endpoint_url"; then
+    endpoint_url="$(bashio::config 'endpoint_url' '')"
+    bashio::log.info "Configure S3 Backup endpoint to ${endpoint_url}"
+    aws configure set s3.endpoint_url "${endpoint_url}"
+    aws configure set s3api.endpoint_url "${endpoint_url}"
+fi
+
+if bashio::config.has_value "max_concurrent_requests"; then
+    aws configure set s3.max_concurrent_requests "$(bashio::config 'max_concurrent_requests')"
```
Comment (thomasfr) [inline s3-backup/run.sh]: Nice addition! Thanks

File: s3-backup/config.json [hunk 2]
```diff
@@ -20,6 +20,8 @@
   "options": {
     "aws_access_key": "",
     "aws_secret_access_key": "",
+    "endpoint_url": "",
+    "max_concurrent_requests": 10,
```
Comment (thomasfr) [inline s3-backup/config.json]: The default value is already 10 (https://docs.aws.amazon.com/cli/latest/topic/s3-config.html#max-concurrent-requests).
Could you make it optional, the same way you are doing it with endpoint_url? This way, one can just leave it as is and be sure that the default value is used by the cli by default.

File: s3-backup/run.sh [hunk 3]
```diff
@@ -1,26 +1,38 @@
...
 delete_local_backups="$(bashio::config 'delete_local_backups' 'true')"
 local_backups_to_keep="$(bashio::config 'local_backups_to_keep' '3')"
 monitor_path="/backup"
 jq_filter=".backups|=sort_by(.date)|.backups|reverse|.[$local_backups_to_keep:]|.[].slug"
 
-export AWS_ACCESS_KEY_ID="$(bashio::config 'aws_access_key')"
-export AWS_SECRET_ACCESS_KEY="$(bashio::config 'aws_secret_access_key')"
-export AWS_REGION="$bucket_region"
+bashio::log.info "Configure S3 Backup..."
+aws configure set aws_access_key_id "$(bashio::config 'aws_access_key')"
```
Comment (thomasfr) [inline s3-backup/run.sh]: I do not think this is necessary. I am not aware that there is a benefit of setting it, especially in this situation since its a one-off script anyways. Could you revert this change please?

File: s3-backup/run.sh [hunk 4]
```diff
@@ -1,26 +1,38 @@
...
+    bashio::log.info "Configure S3 Backup endpoint to ${endpoint_url}"
+    aws configure set s3.endpoint_url "${endpoint_url}"
+    aws configure set s3api.endpoint_url "${endpoint_url}"
+fi
+
+if bashio::config.has_value "max_concurrent_requests"; then
+    aws configure set s3.max_concurrent_requests "$(bashio::config 'max_concurrent_requests')"
+fi
+
+aws configure set region "$(bashio::config 'bucket_region' 'eu-central-1')"
```
Comment (thomasfr) [inline s3-backup/run.sh]: Please do not set region like that, i would like to be as explicit as possible with the cli command as it makes it easier to reason on one line what will happen. Otherwise a reader of the code has to go through the whole command or log outputs.

File: s3-backup/run.sh [hunk 5]
```diff
@@ -1,26 +1,38 @@
...
+bashio::log.info "Configure S3 Backup..."
+aws configure set aws_access_key_id "$(bashio::config 'aws_access_key')"
+aws configure set aws_secret_access_key "$(bashio::config 'aws_secret_access_key')"
+
+aws configure set plugins.endpoint awscli_plugin_endpoint
+if bashio::config.has_value "endpoint_url"; then
+    endpoint_url="$(bashio::config 'endpoint_url' '')"
+    bashio::log.info "Configure S3 Backup endpoint to ${endpoint_url}"
+    aws configure set s3.endpoint_url "${endpoint_url}"
+    aws configure set s3api.endpoint_url "${endpoint_url}"
```
Comment (thomasfr) [inline s3-backup/run.sh]: I would highly prefer having this set as cli parameters. This way we can output pretty much all used parameters on line (https://github.com/thomasfr/hass-addons/pull/5/files#diff-06c5463fab0981ee606b95f4aee978538574d4d7a9884998c868a4f6370273a6R34)

**Review by thomasfr — CHANGES_REQUESTED** [review]
Hey, sorry for my late reply. Had a car accident recently and was off.
Nevertheless, thank you very much for your changes.

In general, i understand why you want to rename it. I do not want to rename though, let me explain why: 
By default it uses Amazon S3, the Amazon S3 APIs, the official aws cli and will work that way by default. Additionally, i personally think its easier to search for and easier to understand what it does - since "S3" is not explicit enough imho.
So please remove all changes where you remove Amazon and AWS.

Besides that, i would like to keep also the cli command as explicit as possible with all used arguments and parameters. In this scenario, i do not see any benefits of setting the parameters first in the awscli configuration.

Thanks again, looking forward for your changes. Will then merge as asap as possible ;)

**Comment by garritfra** [comment]
Really looking forward to seeing this merged. Thanks for your work! :)

**Comment by hobbypunk90** [comment]
sorry for my long delay, i was a little bit busy the last weeks 😒
i hope everything is fine again after your accident.

**Comment by hobbypunk90** [comment]
@thomasfr i reverted the renaming stuff.
I fully understand why you don't like the way i set the configs, i have a similar problem with that.
The problem is, that the endpoint/endpointapi is not able to setup via cli, as i know :see_no_evil: 

This is the reason why i think, it's better to use the `aws configure` way to setup, instead of mix env variables, `aws configure` and cli :joy:

**Comment by bwint** [comment]
Hi there!
Are there any news regarding this pull request?
Would absolutely love to use this repo for backing up my hass instance to a non aws s3 storage.

**Comment by DavidBoettger** [comment]
Hi,
Are there any updates? It seems to me that everything should be working fine. I am really looking forward for this.

**Comment by jpillora** [comment]
Keen to use cloudflare R2 with this PR 😄
